### PR TITLE
Typescript constructor funcs

### PR DESCRIPTION
**Pull request by nokome**
This change attempts to balance the convenience of unnamed arguments, with the robustness of named arguments (aka keyword arguments in Python and other langs) - particularly with respect to changes in the number and order of required properties. 

It should mean that consumers of these factory functions do not need to change their code as often, when/if there are changes to type schemas. For example, with our Python bindings a developer can use keyword arguments to provide some robustness against changes in the number, or order of required args.

```python
h2 = Heading(["Heading Text"], depth = 2)
````

With this change you can do a similar thing in Typescript:

```ts
const h2 = heading(["Heading Text"], { depth: 2 })
````

while not having the burden of having to always specify the name e.g

```ts
// Can still do this
emphasis(['some text'])
// and don't have to do this all the time
emphasis({content: 'some text'})
```

I'm making this change ahead of some proposals for changing some `requireds` in schemas e.g. `depth` in `Heading` (default to 1 instead) and not requiring `author` and `title` for `Article

### CODE REVIEW

File: ts/bindings/typescript.ts [hunk 1]
```diff
@@ -135,17 +135,27 @@ export const typeGenerator = (schema: Schema): string => {
       ({ name, schema }) =>
         `@param ${name} {${schemaToType(schema)}} ${schema.description}`
     ),
```
Comment (alex-ketch) [inline ts/bindings/typescript.ts]: I think this bit isn't quite right any more as it produces a mismatch between the function arity. E.g. `Article`:

``` ts
/**
 * Create a `Article` node
 * @param authors {Array<Person | Organization>} The authors of this creative work.
 * @param title {string | Array<Node>} The title of the creative work.
 * @param other Other properties
 * @returns {Article}
 */
export const article = (
  authors: Array<Person | Organization>,
  props: { title: string | Array<Node> } & OptionalProps<Article>
): Article => ({
  authors,
  ...compact(props),
  type: 'Article'
})
```

Note that the `@param title` is in the Doc Comment, but not part of the function argument.

**Comment by nokome** [comment]
Thanks for the comments @alex-ketch. I agree with your points about consistency. If we all agree that:

> wouldn't be too burdensome for the DX to type constructor({content: […]}

Then I'm happy to go with that approach. @beneboy Your vote?

**Comment by beneboy** [comment]
I don't really have enough knowledge comment.

**Comment by nokome** [comment]
OK, let's go ahead with just a single parameter i.e. `props: Omit<Article, 'type'>` for everything. @alex-ketch could you take this over?

**Comment by alex-ketch** [comment]
@nokome I've rebased this on latest master, so should be good to merge given tests passing.

**Pull request merged**